language/stdlib/modules/association.move

```diff
@@ -109,6 +114,156 @@ module Association {
     acquires PrivilegedCapability {
         Transaction::assert(addr_is_association(addr), 1002);
     }
-}
 
+// **************** SPECIFICATIONS ****************
+// Note: I experimented with using Markdown to organize this, but I haven't converted
+// it to work with the new doc generator, yet.
+
+// # Specifications for association.move
+
+// This is preliminary.  This is one of the first real module libraries with global
+// specifications, and everything is evolving.
+
+    spec module {
+        pragma verify = true;
+
+        // This mirrors the Move function root_address()
+        define spec_root_address(): address { 0xA550C18 }
+
+        // mirrors Move addr_is_association
+        define spec_addr_is_association(addr: address): bool {
+            exists<PrivilegedCapability<T>>(addr)
+            && global<PrivilegedCapability<T>>(addr).is_certified
+        }
+     }
+
+    // ## Management of privileges
+    // TODO: With grant/remove etc., separately specify that only root may grant or remove privileges
+
+    // ### Management of Root {} marker.
+
+    // The root_address is marked by a Root{} object that is stored at that place only.
+    spec module {
+        // enclose this in a function so we can easily repeat the same property in the invariant
+        // and as a post-condition to initialize (the base case).
+        // SPEC: Only the root address has a Root{} resource.
+        define root_invariant(): bool {
+            all(addresses(), |a| exists<Root>(a) ==> (a == spec_root_address()))
+        }
+    }
+    spec schema OnlyRootAddressHasRootPrivilege {
+        invariant root_invariant();
+    }
+    spec module {
+        // apply to all functions except "initialize" and functions that "initialize" calls
+        // before the invariant is established.
+        // Note: All the called functions OBVIOUSLY cannot affect the invariant.
+        // TODO: Try to find a better approach to this that does not require excepting functions.
+        apply OnlyRootAddressHasRootPrivilege to *<T>, *
+            except initialize, root_address, has_privilege, addr_is_association,
```

Comment:
Why do we need to exclude anything else than initialize? While those other functions do not violate the module invariant, it doesn't hurt to apply the module invariant to them.

I would expect that in general, we only exclude the special case of initialize. This gives us peace of mind regards soundness of that the module invariant can be assumed for external calles.

Reply:
Applying the module invariant puts the invariant in as a "requires".  But, when the function (e.g. root_address) is called from initialize, the invariant doesn't hold, and so their is an error reported because of the precondition that doesn't hold.

These functions only need the invariant as a precondition to prove the invariant as a post-condition.  So, they don't really care whether the invariant holds when they are called during initialization.

I think there is a better approach to this problem (I don't think it requires changes to the verifier, just to the specifications).  I will think about it some more.